ft-sdk/src/auth_provider.rs

```diff
@@ -0,0 +1,838 @@
+//! user will be sent to username provider's "create-username" page. If the user wants to log in
+//! via another provider that provides its own username, the username by that provider will be
+//! used if it is available. If the username is not available, the user will be asked to pick a
+//! new username by going to "create-username" page of the provider that provides username, with
+//! the username as default value.
+//!
+//! # How Will Users Update Their Data?
+//!
+//! ft_sdk::auth creates a bunch of functions that can be used to update user data, name, email,
+//! username etc. The UI will have been provided by the auth provider, or some other generic auth
+//! setting package.
+
+/// In the current session, we have zero or more scopes dropped by different auth
+/// providers that have been used so far. Each auth provider sdk also provides some
+/// APIs that require certain scopes to be present. Before calling those APIs, the
+/// caller can check if the session has enough scopes to call that api. If not, the
+/// caller can request the user to log in again with the required scopes.
+pub struct Scope(pub String);
+
+/// update user data
+pub fn authenticate(
+    conn: &mut ft_sdk::Connection,
+    provider_id: &str,
+    // GitHub may use username as Identity, as user can understand their username, but have never
+    // seen their GitHub user id. If we show that user is logged in twice via GitHub, we have to
+    // show some identity against each, and we will use this identity. Identity is mandatory. It
+    // will be stored as UserData::Identity.
+    //
+    // For the same provider_id, if identity changes, we will only keep the latest identity.
+    identity: &str,
+    data: Vec<ft_sdk::auth::UserData>,
+    user_id: Option<ft_sdk::auth::UserId>,
+    // TODO:
+    // _token: Option<serde_json::Value>,
+) -> Result<ft_sdk::UserId, AuthError> {
+    let user_id = if let Some(id) = user_id {
+        id
+    } else {
+        create_empty_user(conn)?
+    };
+
+    modify_user(&user_id, conn, provider_id, identity, data)?;
+
+    Ok(user_id)
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum AuthError {
+    #[error("diesel error: {0}")]
+    Diesel(#[from] diesel::result::Error),
+    #[error("ft_sdk::auth::UserData::Name is required")]
+    NameNotProvided,
+}
+
+/// returns `true` if there's a [UserData::VerifiedEmail] for the provided email
+///
+/// this makes a db call to check if the email is already verified.
+pub fn check_email(
+    conn: &mut ft_sdk::Connection,
+    _email: &str,
+) -> Result<bool, diesel::result::Error> {
+    use db::fastn_user;
+    use diesel::dsl::{count, sql};
+    use diesel::prelude::*;
+    use diesel::sql_types::{Bool, Text};
+
+    #[cfg(not(feature = "postgres"))]
+    let filter = sql::<Bool>("data->'email'->'data'->'verified_emails' LIKE '%\"")
+        .bind::<Text, _>(_email)
+        .sql("\"%'");
```

Comment:
This assumes email is only provided by an auth provider named `email`. We have to loop through all provider data and see if any one of them provided us verified email. Github, Google etc provide verified email as well for example. If you want to check for verified email provided by a given provider only we should add that providers name as an argument to this function. Maybe as `provider: Option<String>` so I can pass None, and am interested in first verified email.